### DISCUSSION TITLE
Tokio interaction with file descriptor limits (RLIMIT_NOFILE)

### DISCUSSION THREAD

**Discussion by Firstyear** [tokio-rs/tokio#5091] [Q&A] [answered]
Hi there,

It's not clear from docs or source code how tokio interacts with platform file descriptor limits. This has an impact on how many sockets can be opened or files opened. Is there some information pertaining to this? 

Thanks!

**Comment by Darksonn** [comment, marked as answer]
You will get an error when trying to create a new file or socket if there are not enough descriptors.

**Reply by Firstyear** [reply]
Okay, but shouldn't there be a way that tokio tries to either report those limits, or allows control of them to say "okay, this many FD's for networking" ?

**Reply by yaleman** [reply]
Alternatively, is there an async way to limit the number of concurrent connections to a given listener?

**Reply by Noah-Kennedy** [reply]
Tokio doesn't manage the file descriptor limit for you. If you are operating a server, you should be tuning your file descriptor limit so that your applications can run under whatever the highest throughput you need to support is. If you are writing a server, you should be handling explicitly the case where accepting a connection fails due to the FD limit.

**Reply by Darksonn** [reply]
You can use a `tokio::sync::Semaphore` to manage access to file descriptors.
